chain/index/api.go

```diff
@@ -0,0 +1,330 @@
+		IndexedEventEntriesCount: uint64(indexedData.nonRevertedEventEntriesCount),
+		Backfilled:               bf,
+	}, nil
+}
+
+func (si *SqliteIndexer) validateIsNullRound(ctx context.Context, epoch abi.ChainEpoch) (*types.IndexValidation, error) {
+	// make sure we do not have tipset(reverted or non-reverted) indexed at this epoch
+	var isNullRound bool
+	err := si.stmts.hasNullRoundAtHeightStmt.QueryRowContext(ctx, epoch).Scan(&isNullRound)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to check if null round exists at height %d: %w", epoch, err)
+	}
+	if !isNullRound {
+		return nil, xerrors.Errorf("index corruption: height %d should be a null round but is not", epoch)
+	}
+
+	return &types.IndexValidation{
+		Height:      uint64(epoch),
+		IsNullRound: true,
+	}, nil
+}
+
+func (si *SqliteIndexer) getTipsetCountsAtHeight(ctx context.Context, height abi.ChainEpoch) (revertedCount, nonRevertedCount int, err error) {
+	err = si.stmts.countTipsetsAtHeightStmt.QueryRowContext(ctx, height).Scan(&revertedCount, &nonRevertedCount)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			// No tipsets found at this height
+			return 0, 0, nil
+		}
+		return 0, 0, xerrors.Errorf("failed to query tipset counts at height %d: %w", height, err)
+	}
+
+	return revertedCount, nonRevertedCount, nil
+}
+
+type indexedTipSetData struct {
+	nonRevertedMessageCount      int
+	nonRevertedEventCount        int
+	nonRevertedEventEntriesCount int
+}
+
+// getIndexedTipSetData fetches the indexed tipset data for a tipset
+func (si *SqliteIndexer) getIndexedTipSetData(ctx context.Context, ts *types.TipSet) (*indexedTipSetData, error) {
+	tsKeyCidBytes, err := toTipsetKeyCidBytes(ts)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	var data indexedTipSetData
+	err = withTx(ctx, si.db, func(tx *sql.Tx) error {
+		if err = tx.Stmt(si.stmts.getNonRevertedTipsetMessageCountStmt).QueryRowContext(ctx, tsKeyCidBytes).Scan(&data.nonRevertedMessageCount); err != nil {
+			return xerrors.Errorf("failed to query non reverted message count: %w", err)
+		}
+
+		if err = tx.Stmt(si.stmts.getNonRevertedTipsetEventCountStmt).QueryRowContext(ctx, tsKeyCidBytes).Scan(&data.nonRevertedEventCount); err != nil {
+			return xerrors.Errorf("failed to query non reverted event count: %w", err)
+		}
+
+		if err = tx.Stmt(si.stmts.getNonRevertedTipsetEventEntriesCountStmt).QueryRowContext(ctx, tsKeyCidBytes).Scan(&data.nonRevertedEventEntriesCount); err != nil {
+			return xerrors.Errorf("failed to query non reverted event entries count: %w", err)
+		}
+
+		return nil
+	})
+
+	return &data, err
+}
+
+// verifyIndexedData verifies that the indexed data for a tipset is correct
+// by comparing the number of messages and events in the chainstore to the number of messages and events indexed.
+// NOTE: Events are loaded from the executed messages of the tipset at the next epoch (ts.Height() + 1).
```

Comment:
```suggestion
//
// Notes:
//
// - Events are loaded from the executed messages of the tipset at the next epoch (ts.Height() + 1).
// - This is not a comprehensive verification because we only compare counts, assuming that a match
//   means that the entries are correct. A future iteration may compare message and event details to
//   confirm that they are what is expected.
```

I want to note this somewhere because it's mildly misleading that verify gives you a full ✅, but that's not quite true. The option I'm outlining is that in the future we could do a more comprehensive check, perhaps that could be enabled with a new boolean argument, or we could take a new argument that allows you to forcibly overwrite regardless of whether it's found correct or not (i.e. "backfill this regardless of what you find").

One thought about a future path to doing this is that we could take the existing boolean argument and creatively parse it in the future to be one of:

- `true` / `false` - existing behaviour
- `"complete-check"` - no backfill, but do a comprehensive check
- `"complete-backfill"` - backfill after a comprehensive check
- `"force-backfill"` - forcibly backfill without checking, just purge and replace

🤷 just thinking out loud, would like to hear your thoughts

Reply:
See https://github.com/filecoin-project/lotus/issues/12570 and https://github.com/filecoin-project/lotus/pull/12571 for exploration of more complete checking.

Reply:
/cc @akaladarshi you might be interested in that too since you've had your head in here

Reply:
@rvagg I agree we should do this. Let's do it in a follow up PR.

Reply:
Filed https://github.com/filecoin-project/lotus/issues/12594 to track this.